lib/integrations/awsoidc/idp_iam_config.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awsoidc
+
+import (
+	"context"
+	"log"
+	"net/url"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	awslib "github.com/gravitational/teleport/lib/cloud/aws"
+)
+
+const (
+	descriptionOIDCIdPRole = "Used by Teleport to provide access to AWS resources."
+)
+
+// IdPIAMConfigureRequest is a request to configure the required Policies to use the EC2 Instance Connect Endpoint feature.
+type IdPIAMConfigureRequest struct {
+	// Cluster is the Teleport Cluster.
+	// Used for tagging the created Roles/IdP.
+	Cluster string
+
+	// AccountID is the AWS Account ID.
+	// Optional. sts.GetCallerIdentity is used if not provided.
+	AccountID string
+
+	// IntegrationName is the Integration Name.
+	// Used for tagging the created Roles/IdP.
+	IntegrationName string
+
+	// Region is the AWS Region.
+	// Used to set up the AWS SDK Client.
+	Region string
+
+	// ProxyPublicAddress is the URL to use as provider URL.
+	// This must be a valid URL (ie, url.Parse'able)
+	// Eg, https://<tenant>.teleport.sh, https://proxy.example.org:443, https://teleport.ec2.aws:3080
+	ProxyPublicAddress string
+
+	// issuer is the above value but only contains the host.
+	// Eg, <tenant>.teleport.sh, proxy.example.org, teleport.ec2.aws:3080
+	issuer string
+
+	// IntegrationRole is the Integration's AWS Role used to set up Teleport as an OIDC IdP.
+	IntegrationRole string
+}
+
+// CheckAndSetDefaults ensures the required fields are present.
+func (r *IdPIAMConfigureRequest) CheckAndSetDefaults() error {
+	if r.Cluster == "" {
+		return trace.BadParameter("cluster is required")
+	}
+
+	if r.IntegrationName == "" {
+		return trace.BadParameter("integration name is required")
+	}
+
+	if r.Region == "" {
+		return trace.BadParameter("region is required")
+	}
+
+	if r.IntegrationRole == "" {
+		return trace.BadParameter("integration role is required")
+	}
+
+	if r.ProxyPublicAddress == "" {
+		return trace.BadParameter("proxy public address is required")
+	}
+
+	issuerURL, err := url.Parse(r.ProxyPublicAddress)
+	if err != nil {
+		return trace.BadParameter("proxy public address is not a valid url: %v", err)
+	}
+	r.issuer = issuerURL.Host
+	if issuerURL.Port() == "443" {
+		r.issuer = issuerURL.Hostname()
```

Comment:
We might have the port in the issuer
Eg `https://proxy.teleport.example.com:3080`
In this case, it will be `proxy.teleport.example.com:3080`
In this case `Hostname()` returns `proxy.teleport.example.com` and we want `proxy.teleport.example.com:3080`

Reply:
Oh got it. I thought the issuer is on the AWS side and will always be 443.